Make several changes to make installation more consistent across platforms and improve usability

WilfS opened the pull request:
This is quite a big pull request! But the changes are relatively basic.

Fixed mac install: 
There was an option for installation on macs, but the installation failed on macOS, mainly because macOS uses .dylib libraries rather than .so libraries. Added an option to install on arm64 architecture too, which most macs are nowadays. Had to sed some changes to ratpac's CMakeLists.txt, which isn't ideal, but can't be fixed permanently without changes to ratpac itself.

Added more packages:
Xerces-C (needed by Geant4) and its dependency ICU are rarely pre-installed on users' machines, so I've added them to the installation. 

Improved error handling:
It was difficult to work out where the install had gone wrong sometimes. I've made it so the install stops as soon as it runs into an error, and it reports back the line where it failed.

Improved bash formatting:
Used shellcheck to reformat setup.sh with recommended bash standard. Mainly adding quotes to variable to prevent expansion.

Commit:
Add installation steps for xerces and icu and change shebang

Xerces and ICU seem to be common missing dependencies, so I've added them into the installation.

The sheband in setup.sh now uses '/usr/bin/env bash', which should get the first instance of bash in the PATH variable. This should avoid problems with multiple installations of bash (a particular problem with homebrew) and is a portable method across systems.

Commit:
Update various installation steps to increase portability

Installs on mac were failing due to using 'dylib' instead of 'so' and a few other things.
Also added an "--arm64" option for building on arm64 architecture. I wasn't able to include tensorflow for this.
Added tests for every install stage.
Ratpac requires some edits to be installed on macs. Here, I've made the edits using sed after ratpac is cloned. This isn't ideal. Hopefully we can make ratpac more universal so these edits aren't needed.

Commit:
Add error handling

setup.sh should now quit as soon as it runs into an error and will print a helpful message saying what part of the install failed.

Commit:
Fix a syntax error

Commit:
Fix enable_mac option

Was using wrong syntax for the if line that checked if we were using a mac

Commit:
Go to correct directory after xerces build

Commit:
Add DYLD_LIBRARY_PATH to env.sh

Commit:
Fix xerces install

Hard-coded dylib as the default library. Now dylib is only used if a mac install is specified.

Commit:
Force xerces to install library in lib, not lib64

Commit:
Build ICU using c++17

Commit:
Fix geant search for xerces

Was using the wrong variable for finding the xerces library

Commit:
Remove full path to cmake

Commit:
Fix removal of ^M from cry Makefile

Commit:
Merge branch 'WilfS/devForMacAndXerces' of github.com:WilfS/ratpac-setup into WilfS/devForMacAndXerces

Commit:
Update readme with new installed packages

Commit:
Fix xerces library variable for ratpac install

Commit:
Improve bash formatting

Used shellcheck to help improve the shell script formatting. Mainly things like adding double quotes around expanded variables and adding '|| exit' to certain commands to catch failures. Should help catch install issues and aid future development

Commit:
Force libzmq to install in lib folder, not lib64

Commit:
Fix mac option in help function

Commit:
Fix more mac options

Commit:
Use \r in sed of cry makefile

Commit:
Adjust LD_LIBRARY_PATH on cry install

This allows the testMain program to be run.

Commit:
Adjust arm64 flags

Commit:
Force geant4 to use lib, not lib64

Commit:
Missed a broken mac flag

Commit:
Change cd behaviour in ratpac install

Commit:
More adjustments to cd behavior

Commit:
Edit mac install

Escaped the $ character in one of the sed commands

Commit:
Change order of install

Print help message and read in arguments before checking for dependencies.

Commit:
Fix some issues cause by install changes

Some variables weren't being set due to reordering the code. Now fixed.

Commit:
Fix some issues cause by install changes

Some variables weren't being set due to reordering the code. Now fixed.

Commit:
Merge branch 'WilfS/devForMacAndXerces' of github.com:WilfS/ratpac-setup into WilfS/devForMacAndXerces

Commit:
Uncomment code

Accidentally left some code commented from tests

Commit:
Change CPPFLAGS to CXXFLAGS

ICU build was complaining -std=c++17 is only for c++ and not c. This should fix it.

Commit:
Update LD_LIBRARY_PATH in cry_install

WilfS committed:
Merge branch 'rat-pac:main' into WilfS/devForMacAndXerces

Commit:
Fix write to env.sh after ratpac_install and edit help message

Commit:
Force nlopt to write libraries to lib and not lib64

Commit:
Fix arm64 and mac options

MarcFBergevin commented:
For my version of Mac `declare -A` is not a recognized option. (`declare [-afFirtx] [-p] [name[=value] ...]`)

MarcFBergevin commented:
For Mac switching to zsh seems to help the process along.

WilfS commented:
Thanks for looking at this, Marc. 
Macs usually use zsh by default, so I've set the mac install up assuming zsh is being used. We could maybe put a note in the readme saying it requires zsh. Could add an option for bash installation in the future too if that's desired.

WilfS commented:
Did switching to zsh also help with the `enable_mac` flag issues you were having?

Commit:
Fix issue on macos where install used high memory

The variable `procuse` that passes on the number of cores to use in the build was being set after options[procuse] was defined. Hence when `make -j"${options[procuse]}"` was being called, it resulted in `make -j` being executed. Not sure why this led to lots of memory usage at the build stage, but setting `procuse` at the correct point seems to have fixed things.

Also updated the docs with a mention of the MacOS installation.

WilfS commented:
I've fixed the issue that were happening with the mac installation. The variable `procuse` was being set too late (I must have moved its initialisation by accident). This led to packages being built with the command `make -j` instead of `make -j8` or equivalent. For whatever reason, this led to the high memory consumption behaviour that was seen. `procuse` is now set correctly and the issue is fixed.

MarcFBergevin commented:
I am having a few issues right now:

- When I try to use the **--only** option, I get `install:34: = not found` error and the install stalls
- `wget` is not installed by default on Macs. So using the alternative `curl -O`, which normally does the trick, does not seem to work as a replacement

Success:
I was able to install **cmake** with the script. But it fails after on the next package to install (**icu**) install.

WilfS committed:
Merge branch 'main' into WilfS/devForMacAndXerces

WilfS commented:
The error you get with the `--only` option is odd. I don't see that behaviour at all. What is the exact command you're using to get that error?
Maybe we should require mac users to install wget. There doesn't seem to be a native equivalent if `curl -O` doesn't work.

MarcFBergevin commented:
Here is the command I was using: `./setup.sh --only cmake root geant4 cry tensorflow ratpac`

I am now trying the full install with the packages pre-dowloaded. 

Let me see if I can pinpoint where it goes wrong with the `=` error.

MarcFBergevin commented:
I have manually download the tar files. The installation of root progressed quite far, but failed after 93%:

`[ 93%] Building CXX object graf3d/ftgl/CMakeFiles/FTGL.dir/src/FTVectoriser.cxx.o
[ 93%] Linking CXX shared library ../../lib/libFTGL.so
ld: warning: ignoring file '/opt/X11/lib/libGL.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
ld: warning: ignoring file '/opt/X11/lib/libGL.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
ld: warning: ignoring file '/opt/X11/lib/libGLU.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
Undefined symbols for architecture arm64:
...
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)`

WilfS commented:
> Here is the command I was using: `./setup.sh --only cmake root geant4 cry tensorflow ratpac`
> 
> I am now trying the full install with the packages pre-dowloaded.
> 
> Let me see if I can pinpoint where it goes wrong with the `=` error.

What OS was this on? I'm guessing not macOS as you don't have the `--mac` flag. I've tested on Red Hat Linux as well and didn't get this issue.

WilfS commented:
> I have manually download the tar files. The installation of root progressed quite far, but failed after 93%:
> 
> ```
> [ 93%] Linking CXX shared library ../../lib/libFTGL.so
> ld: warning: ignoring file '/opt/X11/lib/libGL.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
> ld: warning: ignoring file '/opt/X11/lib/libGL.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
> ld: warning: ignoring file '/opt/X11/lib/libGLU.1.dylib': fat file missing arch 'arm64', file has 'i386,x86_64'
> Undefined symbols for architecture arm64:
> ...
> ld: symbol(s) not found for architecture arm64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)```
> ```

It seems like your install of X11 is built for x86_64 and not arm64. Could you try reinstalling X11 and making sure it's built for arm64? I think you can do it through homebrew...

MarcFBergevin commented:
> > Here is the command I was using: `./setup.sh --only cmake root geant4 cry tensorflow ratpac`
> > I am now trying the full install with the packages pre-dowloaded.
> > Let me see if I can pinpoint where it goes wrong with the `=` error.
> 
> What OS was this on? I'm guessing not macOS as you don't have the `--mac` flag. I've tested on Red Hat Linux as well and didn't get this issue.

I did not realize I needed to add a `--mac` flag. Let me try with that.

WilfS commented:
Ah! That may explain some of the behaviour. You also need the `--arm64` flag if you have a macbook with an M chip.

WilfS committed:
Fix bug introduced by merge with main

hdf5 has been added to the install and I accidentally added a comma when adding it to the install option. Fixed now.

WilfS commented:
I just fixed a bug introduced when the merge with main occurred. Please pull the latest commit!

MarcFBergevin commented:
I have identified the source of the `--only` bug. The option `--` in a if string comparison statement is a bash only option. I have modified the `--` to `-` and it seems to work for me. 

Does this work in bash if you make a similar change?

WilfS commented:
Yes, it does appear to work still in bash.

WilfS commented:
By the way, I'm a bit confused that you're not executing the script with bash. The shebang in setup.sh should ensure the script is executed with bash regardless of the shell it was run in...

MarcFBergevin commented:
> By the way, I'm a bit confused that you're not executing the script with bash. The shebang in setup.sh should ensure the script is executed with bash regardless of the shell it was run in...

Bash version on Mac is very old due to licensing argument for arm64 (if I remember correctly). I believe the bash version dates back to 2007 and is missing some key updates to `declare` such as `declare -A`. `zsh` does not seem to have these issues.

JamesJieranShen commented:
Hi @WilfS , we are reviving some efforts into looking into Mac support, and other things in this PR. Thanks again for submitting this!

A quick question -- have you tried building xerces-c without ICU? As far as I can tell, ICU is a unicode support library and should be very rarely used by xerces-c under geant4, as GDML files do not container unicode characters.

WilfS commented:
Hi @JamesJieranShen, I've not specifically looked into that, no. All I know is that when I tried to install xerces-c it wouldn't work without ICU. There may be some options that allow the build without ICU, in which case I'd happily remove the ICU install. I can look into it this week.

JamesJieranShen commented:
> Hi @JamesJieranShen, I've not specifically looked into that, no. All I know is that when I tried to install xerces-c it wouldn't work without ICU. There may be some options that allow the build without ICU, in which case I'd happily remove the ICU install. I can look into it this week.

[This thread](https://stackoverflow.com/questions/8431659/how-avoid-dependencies-when-using-xerces) suggests that the transcoder can be removed when using `./configure`. I'm not sure if there's a cmake equivalent to that option. ChatGPT seems to think its' `cmake -DXERCES_ENABLE_TRANSCODER=OFF ..`

WilfS commented:
Nice, thanks. I'll try that cmake option.

The main thing I need feedback on regarding the mac install is how to deal with ROOT. The desired "v6-28-00-patches" branch won't build on recent macos versions. I've got around this issue myself by cloning and building the "latest-stable" branch of ROOT, but I imagine that wouldn't be ideal in general—for example minuit is not included in the latest ROOT versions.

WilfS commented:
`-DXERCES_ENABLE_TRANSCODER=OFF` isn't a valid option. It doesn't seem you can disable the transcoder in the xerces build, but you can specify it to use iconv rather than ICU (using `-Dtranscoder=iconv`), which seems to be installed by default on most systems. I've checked and installing xerces this way seems to have no effect on the behaviour of the finished build, so this seems like a good solution

JamesJieranShen commented:
> `-DXERCES_ENABLE_TRANSCODER=OFF` isn't a valid option. It doesn't seem you can disable the transcoder in the xerces build, but you can specify it to use iconv rather than ICU (using `-Dtranscoder=iconv`), which seems to be installed by default on most systems. I've checked and installing xerces this way seems to have no effect on the behaviour of the finished build, so this seems like a good solution

Thanks for looking into this! Guess ChatGPT is not always correct (shocker). I do know that it is possible to build using the make-configure paradigm (`./configure; make`), but if iconv is available almost everywhere, then that's probably ok as well. Thanks again!

JamesJieranShen commented:
> Nice, thanks. I'll try that cmake option.
> 
> The main thing I need feedback on regarding the mac install is how to deal with ROOT. The desired "v6-28-00-patches" branch won't build on recent macos versions. I've got around this issue myself by cloning and building the "latest-stable" branch of ROOT, but I imagine that wouldn't be ideal in general—for example minuit is not included in the latest ROOT versions.

per [this](https://root-forum.cern.ch/t/option-minuit2-is-no-longer-supported-in-root-6-32-06-but-still-present-in-build-instructions/62059/2) ROOT 6.32+ the cmake option for minuit2 was removed as root can no longer be built without minuit2. Since this option was turned on by default in 6.28 as well, we can probably safely remove this flag from the buildscript altogether. 

As for version advancement -- RAT's coupling with ROOT is very minimal, so I would imagine changes from 6.28->6.34 would not effect anything. Obviously, one should test this. In the mean time, perhaps we can provide separate flags for several versions of the script. If all is well `--mac` can probably default to 6.34's patch branch, and the linux version can soon follow.

As a general rule of thumb, though, we shouldn't build against `latest`, as API changes are inevitable when we do something  like that.

Commit:
Update ROOT version for mac install

Older versions of ROOT do not build on the latest macOS versions. Our
solution is to checkout v6-34-00-patches for the mac installation.
v6-28-00-patches is still used for non-mac installs.

WilfS commented:
Ah wait looks like the ICU dependency wasn't resolved. Thought I had done that. Let me try and fix that...

Commit:
Remove ICU dependency

ICU is used as a transcoder by xerces-c, but is not always present by
default on machines. The build now tells xerces-c to use iconv rather
than ICU, which is more likely to be pre-installed.

Commit:
Revert geant4 version to default

Accidentally changed the geant4 version in the last commit. This changes
it back to the version used before.

WilfS commented:
I've removed the ICU dependency but of course another issue has sprung up.

It looks like in the latest macos update Apple Clang was updated to use LLVM 19. This clashes with the zlib version used in geant, so the version of geant we're checking out can't be installed on the latest mac update without installing and pointing to LLVM 16 during the installation. The installation seems to work with the latest geant version (v11.3.2), but I don't know how different that is to the version currently being used (v11.1.2)

WilfS commented:
Hi @JamesJieranShen. xercesc can be installed via homebrew. I'm not sure about other package managers. I mainly included it in the install because it seemed so common for it not to be present when installing ratpac on computing clusters, where installing via package manager isn't possible due to permissions. However I agree that restricting the scope of ratpac-setup is desirable, so I'm happy to remove the xercesc install if that's what is desired. I think some warning of the xercesc dependency and guidance around it should be added to the readme, at least.

JamesJieranShen commented:
> Hi @JamesJieranShen. xercesc can be installed via homebrew. I'm not sure about other package managers. I mainly included it in the install because it seemed so common for it not to be present when installing ratpac on computing clusters, where installing via package manager isn't possible due to permissions. However I agree that restricting the scope of ratpac-setup is desirable, so I'm happy to remove the xercesc install if that's what is desired. I think some warning of the xercesc dependency and guidance around it should be added to the readme, at least.

Thanks! I definitely think then that having the xercesc instruction in the readme/as a warning would be a good compromise. However I think you're right in that this package is special in the sense that it's rarely installed, perhaps we can keep your current implementation but make it such that you have to use ./setup.sh --only xercesc and that it won't be installed by default along with other things. Would that (plus including this behavior in the readme) sound like a reasonable thing to you?

WilfS commented:
Yes, that sounds like a good idea. I'll update things...

WilfS committed:
Merge branch 'main' into WilfS/devForMacAndXerces

Commit:
Remove xerces-c from main install

The number of dependencies in ratpac-setup should be minimised. Xerces-C
is normally present on most machines, so it's been removed from the main
installation with the option of installing with a specific command if
needed. The readme has been updated with instructions.

Commit:
Merge branch 'WilfS/devForMacAndXerces' of github.com:WilfS/ratpac-setup into WilfS/devForMacAndXerces

WilfS commented:
Regarding the geant4 issue on the latest macos, only the latest version of geant (11.3.2) is installable, but it is incompatible with ratpac and the ratpac install will fail. I think the only solution is to use the older xcode tools. I'll look into how feasible that is to automate in a script like this.

WilfS commented:
Turns out, not very feasible. It seems running the software on MacOS 15.4 isn't possible unless we update ratpac to use the latest geant version, or maybe if we patch the old geant, either of which would probably be a big job. I can try and look into how difficult it would be. Let me know if anyone has any insight on this.

MarcFBergevin commented:
I am currently playing with the latest version of Geant4 and RATPAC on Linux. It seems that most of the changes are straightforward and only a few warnings are popping up. It does seem to compile and I can run it.

I will try this install on the Mac, and see if I can additionally open a PR with requires changes in a smart way, so that it will load the correct code depending on the G4 version.

WilfS commented:
That would be terrific, Marc. Thanks!
